### PR TITLE
HIVE-25788: Iceberg CTAS should honor location clause and have correct table properties

### PR DESCRIPTION
**Pull request by marton-bod**

### CODE REVIEW

**Review by szlta — APPROVED** [review]
+1

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergSerDe.java [hunk 1]
```diff
@@ -166,22 +167,33 @@ private void createTableForCTAS(Configuration configuration, Properties serDePro
...
         serDeProperties.getProperty(Catalogs.NAME));
   }
 
-  private void assertNotVectorizedTez(Configuration configuration) {
-    if ("tez".equals(configuration.get("hive.execution.engine")) &&
-        "true".equals(configuration.get("hive.vectorized.execution.enabled"))) {
-      throw new UnsupportedOperationException("Vectorized execution on Tez is currently not supported when using " +
-          "Iceberg tables. Please set hive.vectorized.execution.enabled=false and rerun the query.");
-    }
+  private Properties getCTASTableCreationProperties(Properties serDeProperties) {
```
Comment (pvary) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergSerDe.java:183]: maybe `filterCTASTableCreationProperties`?

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergSerDe.java [hunk 2]
```diff
@@ -166,22 +167,33 @@ private void createTableForCTAS(Configuration configuration, Properties serDePro
       serDeProperties.put(InputFormatConfig.PARTITION_SPEC, PartitionSpecParser.toJson(spec));
     }
 
+    // clean up the properties for table creation (so that internal serde props don't become table props)
+    Properties createProps = getCTASTableCreationProperties(serDeProperties);
```
Comment (pvary) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergSerDe.java:171]: If not reused, why not call it inline?

File: iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergCTAS.java [hunk 3]
```diff
@@ -88,6 +89,35 @@ public void testCTASPartitionedFromHiveTable() throws TException, InterruptedExc
...
+  @Test
+  public void testCTASTblPropsAndLocationClause() throws Exception {
+    Assume.assumeTrue(HiveIcebergSerDe.CTAS_EXCEPTION_MSG, testTableType == TestTables.TestTableType.HIVE_CATALOG);
+
+    shell.executeStatement("CREATE TABLE source (id bigint, name string) PARTITIONED BY (dept string) STORED AS ORC");
+    shell.executeStatement(testTables.getInsertQuery(
+        HiveIcebergStorageHandlerTestUtils.CUSTOMER_RECORDS, TableIdentifier.of("default", "source"), false));
+
+    String location = temp.newFolder().toURI().toString();
+    shell.executeStatement(String.format(
```
Comment (pvary) [inline iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergCTAS.java:101]: Can we add a use-case where the location is provided by the user?

File: ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java [hunk 4]
```diff
@@ -7705,6 +7705,13 @@ protected Operator genFileSinkPlan(String dest, QB qb, Operator input)
       } else {
         tableDescriptor = PlanUtils.getTableDesc(tblDesc, cols, colTypes);
       }
+
+      // if available, set location in table desc properties
+      if (tblDesc != null && tblDesc.getLocation() != null && tableDescriptor != null &&
+          !tableDescriptor.getProperties().containsKey(hive_metastoreConstants.META_TABLE_LOCATION)) {
```
Comment (pvary) [inline ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java:7711]: how does this work with the HMSTranslators? (These are the stuff which makes sure that external/managed table is put into the correct location)

File: iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergCTAS.java (hunk 3 above)
Comment (marton-bod) [reply iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergCTAS.java:101]: This new test contains the user-added location clause: `LOCATION '%s'`
The earlier CTAS tests contain tests without the location clause.

File: ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java (hunk 4 above)
Comment (marton-bod) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java:7711]: If the location clause is present in the query, the translator should not touch that.

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergSerDe.java (hunk 2 above)
Comment (marton-bod) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergSerDe.java:171]: I mainly extracted into a function for readability this time, not reuse. But if you think it's better to inline, I can do it as I don't feel strongly about it

**Review by pvary — APPROVED** [review]

**Pull request merged**